File: script_sigma_vs_peak_number.m
%%%Script para alfa, master

clc; clear; close all; clear all;

% Defaults for this blog post
width = 3;     % Width in inches
height = 3;    % Height in inches
alw = 0.75;    % AxesLineWidth
fsz = 13;      % Fontsize >11.5 <15
lw = 1.5;      % LineWidth
msz = 8;       % MarkerSize
%% %%%%% 1)Load %%%%%%%%%%%%%%%%%%%%%%%%%%

Data = load('6_amplitude_60s_800mV_Bias_gain_gate_standar_histo.txt'); %data
Counts = Data(:,2);
ADC_channels = Data(:,1);

%% %%%%% 2)Load %%%%%%%%%%%%%%%%%%%%%%%%%%

 figure
 bar(ADC_channels, Counts)
 title('Spectra for amplitude 6 of the Led Driver')
 xlabel('ADC channels'); ylabel('Counts')
 grid on

%% %%%%% 3)Calcs %%%%%%%%%%%%%%%%%%%%%%%%%%

%I have to fit all the peaks. The easiest version is to count the peaks by
%hand, and find the indexes, and the fit it, in a similar fashion as what
%I did usually in the bachelor and masters degree

%16 peaks, so this will be hard bro. The idea to choose numbers is two
%above and below after and before the counts starts increasing
%drastically/decreasing

%Initialization of the storing
Sigma = [];         
Delta_sigma = [];
N = [];

%###Peak 1 (around 0)

x_data = ADC_channels(122:130);   
y_data = Counts(122:130);
n = 1;                                      %number of peak


%f(x)=1/(sqrt(2pi)*sigma)*exp(-1/2* ((x-<x>)^2/sigma)
%<x> = mean value and sigma the standar deviation

%The gauss1 fit is a1*exp(-((x-b1)/c1)^2)

ajuste = fit(x_data,y_data,'gauss1');
a = ajuste.a1;
mu = ajuste.b1;
sigma = ajuste.c1/sqrt(2);
ci = confint(ajuste,0.95);  %confidence intervals = coefficients with 95% 
    %confidence bounds this are the coefficientes returned when print 
    %in command windows ajuste1

%Print of the command windows
% ajuste1 = 
%      General model Gauss1:
%      ajuste1(x) =  a1*exp(-((x-b1)/c1)^2)
%      Coefficients (with 95% confidence bounds):
%        a1 =       86.36  (75.63, 97.08)
%        b1 =      0.9616  (-1.323, 3.246)
%        c1 =       22.57  (19.22, 25.93)


delta_sigma=abs(ci(1,3)-ci(2,3))/2*1/sqrt(2); %Error of sigma 1.
    %Note that the error is simply the mean value of the extremal bound.
    %1/sqrt(2) to scalate it to convert it into sigma, since the error is
    %of c

%FWHM_Am=2*sqrt(2*log(2))*sigma1 %MeV
%delta_FWHM_Am=delta_sigma1*2*sqrt(2*log(2))

%Plot of the fit
Gauss = @(x) ajuste.a1*exp(-((x-ajuste.b1)/ajuste.c1).^2);
range = linspace(min(x_data),max(x_data));   %to do the plot of the fit
figure; grid on;
plot(x_data, y_data,'ro-')
hold on
plot(range,Gauss(range))
xlabel('ADC channels')
ylabel('Counts')
legend('data','fit')
title(sprintf('peak %d',n))
%It is awesome, so moving on to the controversial peaks!!

%Storing of the values
Sigma = [Sigma, sigma];
Delta_sigma = [Delta_sigma, delta_sigma];
N = [N, n];

%###Peak 2 ##################

x_data = ADC_channels(158:168); 
y_data = Counts(158:168); 
n = 2;

ajuste = fit(x_data,y_data,'gauss1');
a = ajuste.a1;
mu = ajuste.b1;
sigma = ajuste.c1/sqrt(2);
ci = confint(ajuste,0.95);  

delta_sigma=abs(ci(1,3)-ci(2,3))/2*1/sqrt(2); 

%Plot of the fit
Gauss = @(x) ajuste.a1*exp(-((x-ajuste.b1)/ajuste.c1).^2);
range = linspace(min(x_data),max(x_data));   %to do the plot of the fit
figure; grid on;
plot(x_data, y_data,'ro-')
hold on
plot(range,Gauss(range))
xlabel('ADC channels')
ylabel('Counts')
legend('data','fit')
title(sprintf('peak %d',n))
%It is awesome, so F, this works, and python dont :((

%Storing of the values
Sigma = [Sigma, sigma];
Delta_sigma = [Delta_sigma, delta_sigma];
N = [N, n];

%###Peak 3 ##############

x_data = ADC_channels(192:207); 
y_data = Counts(192:207);
n = 3;

ajuste = fit(x_data,y_data,'gauss1');
a = ajuste.a1;
mu = ajuste.b1;
sigma = ajuste.c1/sqrt(2);
ci = confint(ajuste,0.95);  

delta_sigma=abs(ci(1,3)-ci(2,3))/2*1/sqrt(2); 

%Plot of the fit
Gauss = @(x) ajuste.a1*exp(-((x-ajuste.b1)/ajuste.c1).^2);
range = linspace(min(x_data),max(x_data));   %to do the plot of the fit
figure; grid on;
plot(x_data, y_data,'ro-')
hold on
plot(range,Gauss(range))
xlabel('ADC channels')
ylabel('Counts')
legend('data','fit')
title(sprintf('peak %d',n))
%It is awesome, so F, this works, and python dont :((

%Storing of the values
Sigma = [Sigma, sigma];
Delta_sigma = [Delta_sigma, delta_sigma];
N = [N, n];


%###Peak 4 ###################
n = 4;
x_data = ADC_channels(225:243); 
y_data = Counts(225:243); 


ajuste = fit(x_data,y_data,'gauss1');
a = ajuste.a1;
mu = ajuste.b1;
sigma = ajuste.c1/sqrt(2);
ci = confint(ajuste,0.95); 

delta_sigma=abs(ci(1,3)-ci(2,3))/2*1/sqrt(2); 

%Plot of the fit
Gauss = @(x) ajuste.a1*exp(-((x-ajuste.b1)/ajuste.c1).^2);
range = linspace(min(x_data),max(x_data));   %to do the plot of the fit
figure; grid on;
plot(x_data, y_data,'ro-')
hold on
plot(range,Gauss(range))
xlabel('ADC channels')
ylabel('Counts')
legend('data','fit')
title(sprintf('peak %d',n))

%Storing of the values
Sigma = [Sigma, sigma];
Delta_sigma = [Delta_sigma, delta_sigma];
N = [N, n];

%###Peak 5 ###############
n = 5;
x_data = ADC_channels(263:282); 
y_data = Counts(263:282); 


ajuste = fit(x_data,y_data,'gauss1');
a = ajuste.a1;
mu = ajuste.b1;
sigma = ajuste.c1/sqrt(2);
ci = confint(ajuste,0.95); 

delta_sigma=abs(ci(1,3)-ci(2,3))/2*1/sqrt(2); 

%Plot of the fit
Gauss = @(x) ajuste.a1*exp(-((x-ajuste.b1)/ajuste.c1).^2);
range = linspace(min(x_data),max(x_data));   %to do the plot of the fit
figure; grid on;
plot(x_data, y_data,'ro-')
hold on
plot(range,Gauss(range))
xlabel('ADC channels')
ylabel('Counts')
legend('data','fit')
title(sprintf('peak %d',n))

%Storing of the values
Sigma = [Sigma, sigma];
Delta_sigma = [Delta_sigma, delta_sigma];
N = [N, n];


%###Peak 6 ################
n = 6;
x_data = ADC_channels(299:319); 
y_data = Counts(299:319); 

ajuste = fit(x_data,y_data,'gauss1');
a = ajuste.a1;
mu = ajuste.b1;
sigma = ajuste.c1/sqrt(2);
ci = confint(ajuste,0.95); 

delta_sigma=abs(ci(1,3)-ci(2,3))/2*1/sqrt(2); 

%Plot of the fit
Gauss = @(x) ajuste.a1*exp(-((x-ajuste.b1)/ajuste.c1).^2);
range = linspace(min(x_data),max(x_data));   %to do the plot of the fit
figure; grid on;
plot(x_data, y_data,'ro-')
hold on
plot(range,Gauss(range))
xlabel('ADC channels')
ylabel('Counts')
legend('data','fit')
title(sprintf('peak %d',n))

%Storing of the values
Sigma = [Sigma, sigma];
Delta_sigma = [Delta_sigma, delta_sigma];
N = [N, n];


%###Peak 7##############
n = 7;
x_data = ADC_channels(335:356); 
y_data = Counts(335:356); 

ajuste = fit(x_data,y_data,'gauss1');
a = ajuste.a1;
mu = ajuste.b1;
sigma = ajuste.c1/sqrt(2);
ci = confint(ajuste,0.95); 

delta_sigma=abs(ci(1,3)-ci(2,3))/2*1/sqrt(2); 

%Plot of the fit
Gauss = @(x) ajuste.a1*exp(-((x-ajuste.b1)/ajuste.c1).^2);
range = linspace(min(x_data),max(x_data));   %to do the plot of the fit
figure; grid on;
plot(x_data, y_data,'ro-')
hold on
plot(range,Gauss(range))
xlabel('ADC channels')
ylabel('Counts')
legend('data','fit')
title(sprintf('peak %d',n))

%Storing of the values
Sigma = [Sigma, sigma];
Delta_sigma = [Delta_sigma, delta_sigma];
N = [N, n];


%###Peak 8#############
n = 8;
x_data = ADC_channels(371:388); 
y_data = Counts(371:388); 

ajuste = fit(x_data,y_data,'gauss1');
a = ajuste.a1;
mu = ajuste.b1;
sigma = ajuste.c1/sqrt(2);
ci = confint(ajuste,0.95); 

delta_sigma=abs(ci(1,3)-ci(2,3))/2*1/sqrt(2); 

%Plot of the fit
Gauss = @(x) ajuste.a1*exp(-((x-ajuste.b1)/ajuste.c1).^2);
range = linspace(min(x_data),max(x_data));   %to do the plot of the fit
figure; grid on;
plot(x_data, y_data,'ro-')
hold on
plot(range,Gauss(range))
xlabel('ADC channels')
ylabel('Counts')
legend('data','fit')
title(sprintf('peak %d',n))

%Storing of the values
Sigma = [Sigma, sigma];
Delta_sigma = [Delta_sigma, delta_sigma];
N = [N, n];


%###Peak 9#############
n = 9;
x_data = ADC_channels(407:423); 
y_data = Counts(407:423); 

ajuste = fit(x_data,y_data,'gauss1');
a = ajuste.a1;
mu = ajuste.b1;
sigma = ajuste.c1/sqrt(2);
ci = confint(ajuste,0.95); 

delta_sigma=abs(ci(1,3)-ci(2,3))/2*1/sqrt(2); 

%Plot of the fit
Gauss = @(x) ajuste.a1*exp(-((x-ajuste.b1)/ajuste.c1).^2);
range = linspace(min(x_data),max(x_data));   %to do the plot of the fit
figure; grid on;
plot(x_data, y_data,'ro-')
hold on
plot(range,Gauss(range))
xlabel('ADC channels')
ylabel('Counts')
legend('data','fit')
title(sprintf('peak %d',n))

%Storing of the values
Sigma = [Sigma, sigma];
Delta_sigma = [Delta_sigma, delta_sigma];
N = [N, n];


%###Peak 10#############
n = 10;
x_data = ADC_channels(444:458); 
y_data = Counts(444:458); 

ajuste = fit(x_data,y_data,'gauss1');
a = ajuste.a1;
mu = ajuste.b1;
sigma = ajuste.c1/sqrt(2);
ci = confint(ajuste,0.95); 

delta_sigma=abs(ci(1,3)-ci(2,3))/2*1/sqrt(2); 

%Plot of the fit
Gauss = @(x) ajuste.a1*exp(-((x-ajuste.b1)/ajuste.c1).^2);
range = linspace(min(x_data),max(x_data));   %to do the plot of the fit
figure; grid on;
plot(x_data, y_data,'ro-')
hold on
plot(range,Gauss(range))
xlabel('ADC channels')
ylabel('Counts')
legend('data','fit')
title(sprintf('peak %d',n))

%Storing of the values
Sigma = [Sigma, sigma];
Delta_sigma = [Delta_sigma, delta_sigma];
N = [N, n];


%%%From here, it is hard to see where it is a peak, so will stop now!


%% 3) Plot%%%%%%%%%%%%%%%%%%%
% 
% figure; grid on;
% errorbar(N, Sigma, Delta_sigma,'b*-')
% xlabel('Number of peaks')
% ylabel('\sigma')
% title('Standard deviation vs number of peaks')
% 
% %plot of the variance, sigma^2
% Delta_sigma2 = Delta_sigma*sqrt(2).*Sigma
% figure; grid on;
% errorbar(N, Sigma.^2, Delta_sigma2, 'b*-')
% xlabel('Number of peaks')
% ylabel('\sigma^2')
% title('Variance vs number of peaks')

%The plot is perfect, very similar to the one in the Caen book, so now its
%time to fit it!!!!

%%% 3.1) Linear fit to the data

%y = mx+n

[m, n, r, delta_m, delta_n] = Regresion(N,Sigma);

%%% 3.2) Plot=linear fit

figure; 
errorbar(N, Sigma, Delta_sigma,'r*-','LineWidth', 1.5);
hold on
plot(linspace(min(N),max(N)),m*linspace(min(N),max(N))+n,'LineWidth', 1.5);
xlabel('Peak number')
ylabel('\sigma')
title('Standard deviation vs peak number')
lgd = legend('Data', 'linear fit');
lgd.Location = 'NorthWest';
set(gca,'XLim',[0 max(N)+1]);
set(gca, 'FontSize', fsz, 'LineWidth', alw); %<- Set
%Include fit equation + correlation coefficient
grid on
txt = ['y = ' num2str(m) 'x +' num2str(n) '; r = ' num2str(r)];
text(4,15, txt)     %(2,50) are the x,y position
saveas(gcf, 'sigma_vs_peak_number.png')
%y = mx+n

[m2, n2, r2, delta_m2, delta_n2] = Regresion(N,Sigma.^2);

%%% 3.2) Plot=linear fit

figure; 
errorbar(N, Sigma.^2, Delta_sigma2,'r*-','LineWidth', 1.5);
hold on
plot(linspace(min(N),max(N)),m2*linspace(min(N),max(N))+n2,'LineWidth', 1.5);
xlabel('Peak number')
ylabel('\sigma^2')
title('Variance vs peak number')
lgd = legend('Data', 'linear fit');
lgd.Location = 'NorthWest';
set(gca,'XLim',[0 max(N)+1]);
%set(gca,'YLim',[0 3000]);
set(gca, 'FontSize', fsz, 'LineWidth', alw); %<- Set
%Include fit equation + correlation coefficient
grid on
txt = ['y = ' num2str(m2) 'x +' num2str(n2) '; r = ' num2str(r2)];
text(4,300, txt)     %(2,50) are the x,y position
saveas(gcf, 'sigma2_vs_peak_number.png')


%% 4) Discussion%%%%%%555

%Is weird, CAENs book says sigma goes like sqrt(N), and I found thatit goes
%as N, so my sigma2 is cuadratic, since hes is lineal. However, his graph
%axes are weird, so wil try to mimic them:

figure; 
errorbar(N, Sigma.^2, Delta_sigma2,'r*-','LineWidth', 1.5);
hold on
plot(linspace(min(N),max(N)),m2*linspace(min(N),max(N))+n2,'LineWidth', 1.5);
xlabel('Peak number')
ylabel('\sigma^2')
title('Variance vs peak number')
lgd = legend('Data', 'linear fit');
lgd.Location = 'NorthWest';
set(gca,'XLim',[0 max(N)+1]);
set(gca,'YLim',[0 7000]);
set(gca, 'FontSize', fsz, 'LineWidth', alw); %<- Set
%Include fit equation + correlation coefficient
grid on;
txt = ['y = ' num2str(m2) 'x +' num2str(n2) '; r = ' num2str(r2)];
text(4,300, txt)     %(2,50) are the x,y position
saveas(gcf, 'sigma2_vs_peak_number_CAEN_style.png')

%Not ithas the same scale,  and, as expected, the deviation from the
%linearity are diminished, so now is looks way more linear. So, is CAEN
%(which uses data from MAHAMATSU) fooling us??????????

File: script_sigma_vs_peak_number.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Feb  2 10:28:13 2021

@author: dla

SCRIPT PARA EL EXPERIMENTO DE SIGMA VS PEAK NUMBER DEL A.1.1 DE CAEN

"""

#reset to manually clear all the variables
#clear               #to clear the command windows
#%reset -f          #to clear all the variables without confirmation
#magic('reset -sf')

#######General packages useful##

import matplotlib.pyplot as plt  #for simplicity, to not write matplotlib.pyplot
        #everytime we want to plot something

import scipy 
import numpy as np
    #np contain linspaces as np.linspace(a,b,N)

import sys                   #to import functions from other folders!!
sys.path.insert(0, '/home/dla/Python/Functions_homemade')   
            #path where I have the functions

import Fits
import pandas as pd

######3

#%%
#########################1), Data loading #####################3

data  =pd.read_csv('6_amplitude_60s_800mV_Bias_gain_gate_standar_histo.txt', 
                    sep = '	',names = ['ADC_ch', 'counts'])
        #due to the CAEN's s, the sepparator is strange. If you type ' ', it does
        #not work properly. But if you copy the space in the .txt and paste it here,
        #now it works. That is why below I have kept the other version to read this,
        #just in case :)
        
        
        
############Version to read it without pandas########
# with open('6_amplitude_60s_800mV_Bias_gain_gate_standar_histo.txt') as file_object:
            
#         lines = file_object.readlines()
#         print('the number of lines of the 6ampl file is',len(lines))
        
#         ADC_channel_6ampl = []
#         counts_6ampl = []
#         for i in range(len(lines)):            
#             ADC_channel_6ampl.append(float(lines[i].split()[0])) 
#             counts_6ampl.append(float(lines[i].split()[1]))
#####################


#%%    0.1. Representacion
            

        
plt.figure(figsize=(10,6))  #width, heigh 6.4*4.8 inches by default
plt.bar(data['ADC_ch'],data['counts'], width = data['ADC_ch'][1]-data['ADC_ch'][0])        
plt.title("Spectrum of the LED driver with amplitude 6", fontsize=22)           #title
plt.xlabel("ADC channels", fontsize=14)                        #xlabel
plt.ylabel("Counts", fontsize=14)              #ylabel
# Set size of tick labels.
plt.tick_params(axis='both', labelsize=14)              #size of axis
plt.grid(True) 
plt.xlim(-500,5000)                       #limits of x axis (ojimetro)
#plt.ylim(0,)                            #limits of y axis
plt.savefig('histogram_amplitude_6_60s_800mV_bias_gain_gate_standard.png', format='png')
#

#%%


#############2) Calcs##########################
Amplitude = [x for x in range(10+1)] #turns of the amplitude of the Led driver

#I need to count the number of peaks. How can I do that??????????????????????

####For the moment, by counting:
    
N = 14 #peak number, counting the one in 0, and rejecting the one at 4000




#2@@@@@Gaussian fit@@@
#Source: http://emilygraceripka.com/blog/16



def gaussian(x, Heigh, Mean, Std_dev):
    return Heigh * np.exp(- (x-Mean)**2 / (2 * Std_dev**2)) 
    #

        #this is a gaussian function (more general than normal distribution)
        
        #if using math.exp the fit gives error: 
        #(only size-1 arrays can be converted to Python scalars)
        #with numpy.exp everything fine.       
        #
        #FRIENDSHIP ENDED WITH math.exp(), NOW numpy.exp() IS MY 
        #BEST FRIEND
        #

#So, lets now search the indexes for the fit:
    
#$$$$$$$$$$$$$$$$$$$N = 1$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$
#1st peak (aroudn 0)
n = 1      #peak index

#Creation of the array needed to do the fit

x_data = np.array(data['ADC_ch'][122-1:130-1])       #-1 because python starts at 0
                #and the indexes were found at the .txt reader, that starts in line 1      
y_data = np.array(data['counts'][122-1:130-1])

fit = Fits.Gaussian_fit(x_data,y_data)                      #fit


#Storing of the relevant data, sigma and its error
sigma_stored = np.array([])
delta_sigma_stored = np.array([])
N = np.array([])

sigma_stored = np.append(sigma_stored, fit['sigma'])
delta_sigma_stored = np.append(delta_sigma_stored, fit['\Delta(sigma)'])
N = np.append(N, n)



#$$$$$$$$$$$$$$$$$$$N = 2$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$
n = n +1 #peak index

#Creation of the array needed to do the fit
x_data = np.array(data['ADC_ch'][158-1:168-1])   
y_data = np.array(data['counts'][158-1:168-1])

fit = Fits.Gaussian_fit(x_data,y_data)                      #fit


#Storing of the relevant data, sigma and its error

sigma_stored = np.append(sigma_stored, fit['sigma'])
delta_sigma_stored = np.append(delta_sigma_stored, fit['\Delta(sigma)'])
N = np.append(N, n)


#$$$$$$$$$$$$$$$$$$$N = 3$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$
n = n +1 #peak index

#Creation of the array needed to do the fit
x_data = np.array(data['ADC_ch'][192-1:207-1])   
y_data = np.array(data['counts'][192-1:207-1])

fit = Fits.Gaussian_fit(x_data,y_data)                      #fit


#Storing of the relevant data, sigma and its error

sigma_stored = np.append(sigma_stored, fit['sigma'])
delta_sigma_stored = np.append(delta_sigma_stored, fit['\Delta(sigma)'])
N = np.append(N, n)

#Plot
plt.figure(figsize=(10,6))  #width, heigh 6.4*4.8 inches by default
plt.bar(x_data, y_data, label = 'data', width = x_data[1] - x_data[0], edgecolor="black")                         #original data
plt.plot(x_data, gaussian(x_data, fit['heigh'], fit['mean'], fit['sigma']), 'r--')           #fit
plt.title('Gaussian fit of the peak 3', fontsize=22)                      #title
plt.xlabel("ADC channels", fontsize=14)                                    #xlabel
plt.ylabel("Counts", fontsize=14)                                    #ylabel
plt.legend(['gaussian fit', 'data'], fontsize=16) 
plt.tick_params(axis='both', labelsize=14)                  #size of tick labels  
plt.grid(True)                                              #show grid
#plt.xlim(5.35,5.55)                                         #limits of x 
plt.savefig('gaussian_fit_peak_3.png', format='png')


#$$$$$$$$$$$$$$$$$$$N = 4$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$
n = n +1 #peak index

#Creation of the array needed to do the fit
x_data = np.array(data['ADC_ch'][225-1:243-1])   
y_data = np.array(data['counts'][225-1:243-1])  

fit = Fits.Gaussian_fit(x_data,y_data)                      #fit


#Storing of the relevant data, sigma and its error

sigma_stored = np.append(sigma_stored, fit['sigma'])
delta_sigma_stored = np.append(delta_sigma_stored, fit['\Delta(sigma)'])
N = np.append(N, n)


#$$$$$$$$$$$$$$$$$$$N = 5$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$
n = n +1 #peak index

#Creation of the array needed to do the fit
x_data = np.array(data['ADC_ch'][263-1:282-1])   
y_data = np.array(data['counts'][263-1:282-1]) 

fit = Fits.Gaussian_fit(x_data,y_data)                      #fit


#Storing of the relevant data, sigma and its error

sigma_stored = np.append(sigma_stored, fit['sigma'])
delta_sigma_stored = np.append(delta_sigma_stored, fit['\Delta(sigma)'])
N = np.append(N, n)

#$$$$$$$$$$$$$$$$$$$N = 6$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$
n = n +1 #peak index

#Creation of the array needed to do the fit
x_data = np.array(data['ADC_ch'][299-1:319-1])   
y_data = np.array(data['counts'][299-1:319-1]) 

fit = Fits.Gaussian_fit(x_data,y_data)                      #fit


#Storing of the relevant data, sigma and its error

sigma_stored = np.append(sigma_stored, fit['sigma'])
delta_sigma_stored = np.append(delta_sigma_stored, fit['\Delta(sigma)'])
N = np.append(N, n)

#$$$$$$$$$$$$$$$$$$$N = 7$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$
n = n +1 #peak index

#Creation of the array needed to do the fit
x_data = np.array(data['ADC_ch'][335-1:356-1])   
y_data = np.array(data['counts'][335-1:356-1])   

fit = Fits.Gaussian_fit(x_data,y_data)                      #fit


#Storing of the relevant data, sigma and its error

sigma_stored = np.append(sigma_stored, fit['sigma'])
delta_sigma_stored = np.append(delta_sigma_stored, fit['\Delta(sigma)'])
N = np.append(N, n)


#$$$$$$$$$$$$$$$$$$$N = 8$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$
n = n +1 #peak index

#Creation of the array needed to do the fit
x_data = np.array(data['ADC_ch'][371-1:388-1])   
y_data = np.array(data['counts'][371-1:388-1])   


fit = Fits.Gaussian_fit(x_data,y_data)                      #fit


#Storing of the relevant data, sigma and its error

sigma_stored = np.append(sigma_stored, fit['sigma'])
delta_sigma_stored = np.append(delta_sigma_stored, fit['\Delta(sigma)'])
N = np.append(N, n)


#$$$$$$$$$$$$$$$$$$$N = 9$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$
n = n +1 #peak index

#Creation of the array needed to do the fit
x_data = np.array(data['ADC_ch'][407-1:423-1])   
y_data = np.array(data['counts'][407-1:423-1])   

fit = Fits.Gaussian_fit(x_data,y_data)                      #fit


#Storing of the relevant data, sigma and its error

sigma_stored = np.append(sigma_stored, fit['sigma'])
delta_sigma_stored = np.append(delta_sigma_stored, fit['\Delta(sigma)'])
N = np.append(N, n)



#$$$$$$$$$$$$$$$$$$$N = 10$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$
n = n +1 #peak index

#Creation of the array needed to do the fit
x_data = np.array(data['ADC_ch'][444-1:458-1])   
y_data = np.array(data['counts'][444-1:458-1]) 

fit = Fits.Gaussian_fit(x_data,y_data)                      #fit


#Storing of the relevant data, sigma and its error

sigma_stored = np.append(sigma_stored, fit['sigma'])
delta_sigma_stored = np.append(delta_sigma_stored, fit['\Delta(sigma)'])
N = np.append(N, n)

#%%

#############3) Representation of std##########################

#I have compared to amtlab, and python returns better results, with less error,
#so fuck matlab :)

#3.1.plot of std, sigma

# plt.figure(figsize=(8,5))  #width, heigh 6.4*4.8 inches by default
# plt.errorbar(N, sigma_stored, delta_sigma_stored, fmt='.-b', capsize = 5)
# plt.title('Standar deviation vs peak number', fontsize=22)          #title
# plt.xlabel("Peak number ", fontsize=14)                                    #xlabel
# plt.ylabel(r'$\sigma$', fontsize=14)                                    #ylabel
# plt.tick_params(axis='both', labelsize=14)            #size of tick labels  
# plt.grid(True)                                              #show grid
# #plt.xlim(5.35,5.55) 
# plt.savefig('sigma_vs_peaknumber_py.png', format='png')


#3.2.plot of variance, sigma^2

sigma2_stored = sigma_stored**2                    #sigma^2
delta_sigma2_stored = sigma_stored*np.sqrt(2) * delta_sigma_stored
    #this is the way to compute sigma*sqrt(2)*delta_sigma


# plt.figure(figsize=(8,5))  #width, heigh 6.4*4.8 inches by default
# plt.errorbar(N, sigma2_stored, delta_sigma2_stored, fmt='.-b', capsize = 5)
# plt.title('Variance vs peak number', fontsize=22)          #title
# plt.xlabel("Peak number ", fontsize=14)                                    #xlabel
# plt.ylabel(r'$\sigma^2$', fontsize=14)                                    #ylabel
# plt.tick_params(axis='both', labelsize=14)            #size of tick labels  
# plt.grid(True)                                              #show grid
# #plt.xlim(5.35,5.55) 
# plt.savefig('sigma2_vs_peaknumber_py.png', format='png')


# 3.3. Linear fit of both
#Could use my script or scipy, defining a linear function. Since I do not know how
#to compute r from scipy, and the formula used (derived from the formulas used in my
#script) will continue to use my script :))

def linear(x, m, n):       #Definition of the function to use to fit the data
    return m * x + n 


ajuste = Fits.LinearRegression(N, sigma_stored)
            #the results are the same as in my script, so awesome
            
ajuste2 = Fits.LinearRegression(N, sigma2_stored)            

#Sigma plot with fit

plt.figure(figsize=(10,6))  #width, heigh 6.4*4.8 inches by default
plt.errorbar(N, sigma_stored, delta_sigma_stored, fmt='ro', capsize = 10)
plt.plot(N, [linear(a, ajuste['Slope'], ajuste['Intercept']) for a in N], linewidth=3)      #fit
plt.title('Standard deviation vs peak number', fontsize=22)          #title
plt.xlabel("Peak number ", fontsize=14)                                    #xlabel
plt.ylabel(r'$\sigma$', fontsize=14)                                    #ylabel
plt.tick_params(axis='both', labelsize=14)            #size of tick labels  
plt.grid(True)                                              #show grid
plt.legend(['linear fit','data',], fontsize=16, loc='upper left')             #legend
plt.text(6,15, 'y(x) = {0:1.3f}x + {1:1.3f} ; r = {2:1.3f}'
         .format(ajuste['Slope'],ajuste['Intercept'],ajuste['r']), fontsize=14)    #first 2 arguments are x,y position.
    #0:1.3f: 0 is 1st argument in format, 1.3f means float on 3 decimals
plt.xlim(0,15)
plt.savefig('sigma_vs_peaknumber_py.png', format='png')


#Sigma2 plot with fit
plt.figure(figsize=(10,6))  #width, heigh 6.4*4.8 inches by default
plt.errorbar(N, sigma2_stored, delta_sigma2_stored, fmt='ro', capsize = 10)
plt.plot(N, [linear(a, ajuste2['Slope'], ajuste2['Intercept']) for a in N], linewidth=3)      #fit

plt.title('Variance vs peak number', fontsize=22)          #title
plt.xlabel("Peak number ", fontsize=14)                                    #xlabel
plt.ylabel(r'$\sigma^2$', fontsize=14)                                    #ylabel
plt.tick_params(axis='both', labelsize=14)            #size of tick labels  
plt.grid(True)                                              #show grid
plt.legend(['linear fit','data',], fontsize=16, loc='upper left')             #legend
plt.text(5,100, 'y(x) = {0:1.3f}x + {1:1.3f} ; r = {2:1.2f}'
         .format(ajuste2['Slope'],ajuste2['Intercept'],ajuste2['r']), fontsize=14)
plt.xlim(0,15)
plt.savefig('sigma2_vs_peaknumber_py.png', format='png')


#Sigma2 plot with fit, CAENs style
plt.figure(figsize=(10,6))  #width, heigh 6.4*4.8 inches by default
plt.errorbar(N, sigma2_stored, delta_sigma2_stored, fmt='ro', capsize = 10)
plt.plot(N, [linear(a, ajuste2['Slope'], ajuste2['Intercept']) for a in N], linewidth=3)      #fit

plt.title('Variance vs peak number', fontsize=22)          #title
plt.xlabel("Peak number ", fontsize=14)                                    #xlabel
plt.ylabel(r'$\sigma^2$', fontsize=14)                                    #ylabel
plt.tick_params(axis='both', labelsize=14)            #size of tick labels  
plt.grid(True)                                              #show grid
plt.legend(['linear fit','data',], fontsize=16, loc='upper left')             #legend
plt.text(5,200, 'y(x) = {0:1.3f}x + {1:1.3f} ; r = {2:1.2f}'
         .format(ajuste2['Slope'],ajuste2['Intercept'],ajuste2['r']), fontsize=14) #10 default size)
plt.ylim(0,7000) 
plt.xlim(0,15)
plt.savefig('sigma2_vs_peaknumber_CAENs_style_py.png', format='png')




#%% 

########################4)Square fit to variance##########################

#the fashion to do the fit is the usual, define the funciton, and then fit:
    

#Creation of the array needed to do the fit
x_data = np.array(N)   
y_data = np.array(sigma2_stored) 

quad_fit = Fits.QuadraticRegression(x_data, y_data)

def cuadratic(x, a, b, c):       #Definition of the function to use to fit the data
    return a * x**2 + b*x + c 


#plot
plt.figure(figsize=(10,6))  #width, heigh 6.4*4.8 inches by default
plt.errorbar(N, sigma2_stored, delta_sigma2_stored, fmt='ro', capsize = 10)
plt.plot(N, [cuadratic(x, quad_fit['a'], quad_fit['b'], quad_fit['c']) for x in N], linewidth=3)      #fit

plt.title('Variance vs peak number', fontsize=22)          #title
plt.xlabel("Peak number ", fontsize=14)                                    #xlabel
plt.ylabel(r'$\sigma^2$', fontsize=14)                                    #ylabel
plt.tick_params(axis='both', labelsize=14)            #size of tick labels  
plt.grid(True)                                              #show grid
plt.legend(['quadratic fit','data',], fontsize=16, loc='upper left')             #legend
plt.text(2.7,300, 'y(x) = {0:1.3f}x^2 + {1:1.3f}x + {2:1.3f} ; r = {3:1.3f}'
         .format(quad_fit['a'], quad_fit['b'], quad_fit['c'], quad_fit['r']), fontsize=14) #10 default size
plt.xlim(0,15)
plt.savefig('sigma2_vs_peaknumber_cuadratic_fit_py.png', format='png')


#dude, that fit is sick, which would also confirm that CAEN is wrong.